p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,275 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ tpt.CapableConn = &connection{}
+
+const (
+	maxAcceptQueueLen int = 10
+)
+
+type connection struct {
+	pc        *webrtc.PeerConnection
+	transport *WebRTCTransport
+	scope     network.ConnManagementScope
+
+	localPeer      peer.ID
+	privKey        ic.PrivKey
+	localMultiaddr ma.Multiaddr
+
+	remotePeer      peer.ID
+	remoteKey       ic.PubKey
+	remoteMultiaddr ma.Multiaddr
+
+	m       sync.Mutex
+	streams map[uint16]*dataChannel
+
+	acceptQueue chan network.MuxedStream
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newConnection(
+	pc *webrtc.PeerConnection,
+	transport *WebRTCTransport,
+	scope network.ConnManagementScope,
+
+	localPeer peer.ID,
+	privKey ic.PrivKey,
+	localMultiaddr ma.Multiaddr,
+
+	remotePeer peer.ID,
+	remoteKey ic.PubKey,
+	remoteMultiaddr ma.Multiaddr,
+) *connection {
+	acceptQueue := make(chan network.MuxedStream, maxAcceptQueueLen)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	conn := &connection{
+		pc:        pc,
+		transport: transport,
+		scope:     scope,
+
+		localPeer:      localPeer,
+		privKey:        privKey,
+		localMultiaddr: localMultiaddr,
+
+		remotePeer:      remotePeer,
+		remoteKey:       remoteKey,
+		remoteMultiaddr: remoteMultiaddr,
+		ctx:             ctx,
+		cancel:          cancel,
+		streams:         make(map[uint16]*dataChannel),
+
+		acceptQueue: acceptQueue,
+	}
+
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
```

Comment:
FYI I would turn these anonymous functions into private methods, given it is dealing with values that you are assigning to `*connection`.

Reply:
I have turned the `OnConnectionStateChange` to a private method, however this seems to not execute the `OnOpen` callback when I use a private method. This is strange since the callback should execute as soon as the callback is set because a) the channel is remote, and b) the channel is detached. 
refer: https://github.com/pion/webrtc/blob/602cedd4fe8d7a49c4d3526c720d0be3056af521/datachannel.go#L272
and https://github.com/pion/webrtc/blob/602cedd4fe8d7a49c4d3526c720d0be3056af521/datachannel.go#L218

Reply:
I would have to learn a lot more about the codebase and business logic to go deeper into your question, for which I now have no priority AFAIK. But the fact of using a method vs an anonymous function object should make no difference to whether or not it gets executed, so there must be something else wrong in that case.